modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpChannel.java

```diff
@@ -0,0 +1,260 @@
+     * @param request               The request that is handled by this channel.
+     * @param pipelinedRequest      If HTTP pipelining is enabled provide the corresponding pipelined request. May be null if
+     *                              HTTP pipelining is disabled.
+     * @param detailedErrorsEnabled true iff error messages should include stack traces.
+     */
+    Netty4HttpChannel(
+        final Netty4HttpServerTransport transport,
+        final Netty4HttpRequest request,
+        final HttpPipelinedRequest pipelinedRequest,
+        final boolean detailedErrorsEnabled) {
+        super(request, detailedErrorsEnabled);
+        this.transport = transport;
+        this.channel = request.getChannel();
+        this.nettyRequest = request.request();
+        this.pipelinedRequest = pipelinedRequest;
+    }
+
+    @Override
+    public BytesStreamOutput newBytesOutput() {
+        return new ReleasableBytesStreamOutput(transport.bigArrays);
+    }
+
+
+    @Override
+    public void sendResponse(RestResponse response) {
+        // if the response object was created upstream, then use it;
+        // otherwise, create a new one
+        ByteBuf buffer = Netty4Utils.toByteBuf(response.content());
+        FullHttpResponse resp = newResponse(buffer);
+        resp.setStatus(getStatus(response.status()));
+
+        Netty4CorsHandler.setCorsResponseHeaders(nettyRequest, resp, transport.getCorsConfig());
+
+        String opaque = nettyRequest.headers().get("X-Opaque-Id");
+        if (opaque != null) {
+            setHeaderField(resp, "X-Opaque-Id", opaque);
+        }
+
+        // Add all custom headers
+        addCustomHeaders(response, resp);
+
+        BytesReference content = response.content();
+        boolean addedReleaseListener = false;
+        try {
+            // If our response doesn't specify a content-type header, set one
+            setHeaderField(resp, HttpHeaderNames.CONTENT_TYPE.toString(), response.contentType(), false);
+            // If our response has no content-length, calculate and set one
+            setHeaderField(resp, HttpHeaderNames.CONTENT_LENGTH.toString(), String.valueOf(buffer.readableBytes()), false);
+
+            addCookies(resp);
+
+            final ChannelPromise promise = channel.newPromise();
+
+            if (content instanceof Releasable) {
+                promise.addListener(new Netty4ReleaseChannelFutureListener((Releasable) content));
+                addedReleaseListener = true;
+            }
+
+            if (isCloseConnection()) {
+                promise.addListener(ChannelFutureListener.CLOSE);
+            }
+
+            if (pipelinedRequest != null) {
+                channel.writeAndFlush(pipelinedRequest.createHttpResponse(resp, promise));
+            } else {
+                channel.writeAndFlush(resp, promise);
+            }
+
+        } finally {
+            if (!addedReleaseListener && content instanceof Releasable) {
```

Comment:
nit: maybe use a boolean `release = (content instanceof Releasable)` which you set to false if the listener is added. then you just need to check for `if (release)` in the finally block?